core/src/main/scala/org/apache/spark/sql/sources/SparkOptimizations.scala

```diff
@@ -0,0 +1,410 @@
+            case (plan, replaceWith) if plan.fastEquals(q)
+                && !plan.fastEquals(replaceWith) => replaceWith
+          }.reduceLeftOption((acc, p_) => if (!p_.fastEquals(acc)) p_ else acc)
+
+          replacement.map(newP => {
+            val newAttributes = newP.schema.toAttributes
+            q.output.foreach({
+              case f: AttributeReference =>
+                newAttributesMap(f) = newAttributes.find(_.name.equalsIgnoreCase(f.name)).
+                    getOrElse(throw new IllegalStateException(
+                      s"Field $f not found in ${newAttributes}"))
+            })
+
+            newP
+          }).getOrElse(q)
+      }
+
+      if (newAttributesMap.nonEmpty) {
+        newPlan transformUp {
+          case q: LogicalPlan =>
+            q transformExpressionsUp {
+              case a: AttributeReference => newAttributesMap.getOrElse(a, a)
+            }
+        }
+      } else {
+        newPlan
+      }
+    }
+  }
+
+  private def fetchIndexes(table: LogicalPlan) = table.collect {
+    case l@LogicalRelation(p: ParentRelation, _, _) =>
+      val catalog = snappySession.sessionCatalog
+      (l.asInstanceOf[LogicalPlan], p.getDependents(catalog).map(idx =>
+        catalog.lookupRelation(catalog.newQualifiedTableName(idx))))
+    case s@SubqueryAlias(alias, child@LogicalRelation(p: ParentRelation, _, _)) =>
+      val catalog = snappySession.sessionCatalog
+      (s.asInstanceOf[LogicalPlan], p.getDependents(catalog).map(idx =>
+        catalog.lookupRelation(catalog.newQualifiedTableName(idx))))
+  }
+
+  object HasColocatedEntities {
+
+    type ReturnType = (
+        Seq[(BaseColumnFormatRelation, BaseColumnFormatRelation)], // left & right relation
+            Seq[(LogicalPlan, LogicalPlan)], // Seq[left (tablePlan, indexPlan)]
+            Seq[(LogicalPlan, LogicalPlan)] // Seq[right (tablePlan, indexPlan)]
+        )
+
+    def unapply(tables: (LogicalPlan, LogicalPlan)):
+    Option[ReturnType] = {
+      val (left, right) = tables
+
+      /** now doing a one-to-one mapping of lefttable and its indexes with
+        * right table and its indexes. Following example explains the combination
+        * generator.
+        *
+        * val l = Seq(1, 2, 3)
+        * val r = Seq(4, 5)
+        * l.zip(Seq.fill(l.length)(r)).flatMap {
+        * case (leftElement, rightList) => rightList.flatMap { e =>
+        * Seq((leftElement, e))
+        * }
+        * }.foreach(println)
+        * will output :
+        * (1, 4)
+        * (1, 5)
+        * (2, 4)
+        * (2, 5)
+        * (3, 4)
+        * (3, 5)
+        *
+        */
+      val leftRightEntityMapping = for {
+        (leftTable, leftIndexes) <- fetchIndexes(left)
+        (rightTable, rightIndexes) <- fetchIndexes(right)
+        leftSeq = Seq(leftTable) ++ leftIndexes
+        rightSeq = Seq(rightTable) ++ rightIndexes
+      } yield {
+        leftSeq.zip(Seq.fill(leftSeq.length)(rightSeq)) flatMap {
```

Comment:
This is generating the search space for 2 tables and its indexes.
example: tab1 having 2 indexes defined and tab2 is having 3 indexes defined. as shown in the commented code it will generate all combinatorial pairs b/w the tables & indexes. So, this will generate a seq of pairs viz. (t1, t2), (t1, t2_i1), (t1, t2_i2), (t1, t2_i3), (t1_i1, t2), (t1_i2, t2). Rest of the code just eliminates infeasible join relations and comes up with colocated join pairs.